### PR TITLE
Fix filter evaluation on CHAR(n) dictionary columns in SliceDictionarySelectiveReader

### PR DESCRIPTION
**Pull request by yingsu00**
Fixes https://github.com/prestodb/presto/issues/19138

When a filter for a column of CHAR(n) type is pushed down to the stream readers, the padded spaces would be removed. This is different than the filter on VARCHAR(n), where the trailing spaces would be preserved. 
For example, `c = 'AIR         '` would be come `c = 'AIR'` when `SliceDictionarySelectiveReader` takes it at construction time. However the existing code still thinks the filter is with paddings therefore causing the bug in https://github.com/prestodb/presto/issues/19138. When testing a value's length while evaluating the filter on CHAR(n), it is incorrect to use the `maxCodePointCount`(n in CHAR(n)) as the value's length. Instead, the filter's length shall be compared with the value's real length, which was already calculated and stored in `dictionaryOffsetVector`. The tests are updated to take filters with paddings removed.

```
== RELEASE NOTES ==
Hive Changes
* Fix a bug where the filter on a CHAR(n) column is not evaluated correctly for ORC/DWRF files when filter pushdown is enabled.
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java [hunk 1]
```diff
@@ -237,13 +237,8 @@ private int readWithFilter(int[] positions, int positionCount)
...
                 int rawIndex = toIntExact(dataStream.next());
                 int index = inRowDictionary ? stripeDictionarySize + rawIndex : rawIndex;
-                int length;
-                if (isCharType) {
-                    length = maxCodePointCount;
-                }
-                else {
-                    length = inRowDictionary ? rowGroupDictionaryLength[rawIndex] : stripeDictionaryLength[rawIndex];
-                }
+                int length = dictionaryOffsetVector[index + 1] - dictionaryOffsetVector[index];
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java:240]: @sdruzkin  this is the main change for this PR, What kind of testing is required to verify it ? Does unit test capture the changes

**Review by arunthirupathi — COMMENTED** [review]
The code change looks good to me, but will wait for @sdruzkin  to confirm as he has more knowledge about this code.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveReaderContext.java [hunk 2]
```diff
@@ -76,7 +76,7 @@ public Type getOutputType()
     }
 
     @Nullable
-    public TupleDomainFilter getRowGroupFilter(BooleanInputStream presentStream)
+    public TupleDomainFilter getFilter(BooleanInputStream presentStream)
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveReaderContext.java:79]: Was this renaming necessary for this fix?

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 3]
```diff
@@ -361,10 +361,29 @@ public void testRoundTrip(Type type, List<?> readValues, TupleDomainFilter... fi
...
+     * @throws Exception
+     */
+    public void testRoundTrip(
+            Type type,
+            List<?> readValues,
+            List<Map<Subfield, TupleDomainFilter>> columnFilters,
+            List<Map<Subfield, TupleDomainFilter>> valuesFilters)
+            throws Exception
+    {
+        List<Map<Integer, Map<Subfield, TupleDomainFilter>>> streamReaderFilters = columnFilters.stream().map(filter -> ImmutableMap.of(0, filter)).collect(toImmutableList());
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:382]: Please reformat all stream operations and put each stream operation on a new line like this:
```
collection.stream()
  .op()
  .op()
```

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 4]
```diff
@@ -522,10 +542,21 @@ private void testRoundTripType(Type type, List<?> readValues)
...
+            throws Exception
+    {
+        testRoundTripTypes(types, readValues, filters, valuesFilters, ImmutableList.of());
+    }
+
+    public void testRoundTripTypes(
+            List<Type> types,
+            List<List<?>> readValues,
+            List<Map<Integer, Map<Subfield, TupleDomainFilter>>> filters,
+            List<Map<Integer, Map<Subfield, TupleDomainFilter>>> valuesfilters,
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:558]: Rename valuesfilters to valuesFilters

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 5]
```diff
@@ -584,52 +615,70 @@ private List<List<?>> insertNulls(List<Type> types, List<List<?>> values)
...
+            List<Map<Integer, Map<Subfield, TupleDomainFilter>>> filters,
+            List<Map<Integer, Map<Subfield, TupleDomainFilter>>> valuesFilters)
             throws Exception
     {
-        List<OrcReaderSettings> settings = filters.stream()
-                .map(entry -> OrcReaderSettings.builder().setColumnFilters(entry).build())
-                .collect(toImmutableList());
+        ImmutableList.Builder<OrcReaderSettings> settingsBuilder = ImmutableList.builder();
+        for (int i = 0; i < filters.size(); i++) {
+            settingsBuilder.add(OrcReaderSettings.builder().setColumnFilters(filters.get(i)).setExpectedValuesFilters(valuesFilters.get(i)).build());
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:630]: reformat this, extract a new variable for OrcReaderSettings and put each builder operation on a new line

**Review by sdruzkin — COMMENTED** [review]
Can we add an integration test case (the one executing a query) checking it end to end with both dictionaries and without dictionaries? If not, let's not do it, because it will make hotfixing hard.

**Review by arunthirupathi — APPROVED** [review]
Once you address @sdruzkin's comment it is safe to merge.

**Comment by yingsu00** [comment]
> Can we add an integration test case (the one executing a query) checking it end to end with both dictionaries and without dictionaries? If not, let's not do it, because it will make hotfixing hard.

Sure, added this test in AbstractTestDistributedQueries, which would be executed with different configurations. The "batch" readers would be tested in its subclass `TestHiveDistributedQueries`, and "selective" readers  tested in `TestHivePushdownDistributedQueries`.
```
    @Test
    public void testStringFilters()
    {
        assertUpdate("CREATE TABLE test_charn_filter (shipmode CHAR(10))");
        assertTrue(getQueryRunner().tableExists(getSession(), "test_charn_filter"));
        assertTableColumnNames("test_charn_filter", "shipmode");
        assertUpdate("INSERT INTO test_charn_filter SELECT shipmode FROM lineitem", 60175);

        assertQuery("SELECT count(*) FROM test_charn_filter WHERE shipmode = 'AIR'", "VALUES (8491)");
        assertQuery("SELECT count(*) FROM test_charn_filter WHERE shipmode = 'AIR    '", "VALUES (8491)");
        assertQuery("SELECT count(*) FROM test_charn_filter WHERE shipmode = 'AIR       '", "VALUES (8491)");
        assertQuery("SELECT count(*) FROM test_charn_filter WHERE shipmode = 'AIR            '", "VALUES (8491)");

        assertUpdate("CREATE TABLE test_varcharn_filter (shipmode VARCHAR(10))");
        assertTrue(getQueryRunner().tableExists(getSession(), "test_varcharn_filter"));
        assertTableColumnNames("test_varcharn_filter", "shipmode");
        assertUpdate("INSERT INTO test_varcharn_filter SELECT shipmode FROM lineitem", 60175);

        assertQuery("SELECT count(*) FROM test_varcharn_filter WHERE shipmode = 'AIR'", "VALUES (8491)");
        assertQuery("SELECT count(*) FROM test_varcharn_filter WHERE shipmode = 'AIR    '", "VALUES (0)");
        assertQuery("SELECT count(*) FROM test_varcharn_filter WHERE shipmode = 'AIR       '", "VALUES (0)");
        assertQuery("SELECT count(*) FROM test_varcharn_filter WHERE shipmode = 'AIR            '", "VALUES (0)");
    }
```

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java (hunk 5 above)
Comment (yingsu00) [reply presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:630]: Sure, reformatted to
```
        ImmutableList.Builder<OrcReaderSettings> settingsBuilder = ImmutableList.builder();
        OrcReaderSettings.Builder orcSettingBuilder = OrcReaderSettings.builder();
        for (int i = 0; i < filters.size(); i++) {
            settingsBuilder.add(orcSettingBuilder
                    .setColumnFilters(filters.get(i))
                    .setExpectedValuesFilters(valuesFilters.get(i))
                    .build());
        }
```

**Review by sdruzkin — APPROVED** [review]
LGTM

**Comment by sdruzkin** [comment]
@yingsu00 is it good to merge?

**Pull request merged**